pkgs/build-support/trivial-builders.nix

```diff
@@ -761,6 +761,36 @@ rec {
         writeDirectReferencesToFile (writeText "string-file" string);
 
 
+  /**
+   * Write the references (i.e. the runtime dependencies in the Nix store) of the elements in `paths' to a file.
+   * The elements are sorted according to their hashes, and duplicated lines are removed.
+   */
+  writeMultipleReferencesToFile = paths: runCommand "runtime-deps-multiple" {
+    # Uniquely sort the input paths to reduce unnessesary rebuilds
+    referencesFiles = map writeReferencesToFile (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]);
```

Comment:
Ok, so now I'm looking at https://github.com/NixOS/nixpkgs/pull/224636/files#diff-35d95fd2d33596cbcc6f411406cd7a1a847e1d86e206c7aa248732f2254b52d0R38 and I'm beginning to understand what this is meant to be used for. From what I understand, each evaluation of `writeReferencesToFile` amounts to a single `runCommand`.

Question: why should `map writeReferencesToFile (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ])` ($N$ lighter runCommands) be faster than one bigger runCommand:

```nix
let
  directDeps = (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]);
  directDepsTxt = writeText "direct-deps.txt" (concatStringsSep "\n" uniqueDirectDependencies;
in writeReferencesToFile directDepsTxt;
```

Reply:
> Question: why should map writeReferencesToFile (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]) ($N$ lighter runCommands) be faster than one bigger runCommand:\n\nlet\n  directDeps = (lib.pipe (map toString paths) [ (builtins.sort builtins.lessThan) lib.unique ]);\n  directDepsTxt = writeText \"direct-deps.txt\" (concatStringsSep \"\\n\" uniqueDirectDependencies;\nin writeReferencesToFile directDepsTxt;

Cool! It would be perfect if we could remove the line of the direct-list packge store path .

Reply:
Oh, do you mean the txt?

Reply:
Yes.

Reply:
Nothing a little `sed` can't do 😅 I'm more concerned about having numbers on whether doing any of this is worth the effort at all. I thought about it for a moment and realized that benchmarking can be quite a bit of a job: there's nix evaluation cache, there's garbage collection...

Reply:
My workflow to benchmark the build of a package:

Suppose the package to be benchmarked is `hello`, a non-fixed-output derivation in Nixpkgs:

1.  Preparation
    ```sh
    # Reduce the possibility for the target package to appear in the store.
    # should we already have it in the store, the benchmark won't work.
    nix-store --gc
    cd path/to/nixpkgs
    # Get all the dependencies ready.
    nix-shell -A hello --run "sh -c :"
    ```

2.  Actual build. Add timing code around this commnd:
    ```
    nix-build --offline --no-out-link -A hello
    ```

Reply:
Just implemented like that.